core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -184,24 +184,32 @@ public void respond(ActionListener<TransportResponse.Empty> listener) {
     private static boolean finishWrite(IndexShard indexShard, ReplicatedWriteRequest<?> request, Translog.Location location,
             Consumer<Boolean> refreshListener) {
         boolean fsyncTranslog = indexShard.getTranslogDurability() == Translog.Durability.REQUEST && location != null;
-        if (fsyncTranslog) {
-            indexShard.sync(location);
-        }
-        indexShard.maybeFlush();
-        boolean async = false;
+        boolean refreshPending = false;
         switch (request.getRefreshPolicy()) {
         case IMMEDIATE:
             indexShard.refresh("refresh_flag_index");
             break;
         case WAIT_UNTIL:
             if (location != null) {
-                async = true;
-                indexShard.addRefreshListener(location, refreshListener);
+                refreshPending = true;
+                indexShard.addRefreshListener(location, forcedRefresh -> {
+                    if (fsyncTranslog) {
```

Comment:
what I mean is just do it at the end of the method so waiting for refresh will be concurrent to the fsync (just a minor optimization - not really something worth complicating flow paths)